music/src/core/soundfont.ts

```diff
@@ -0,0 +1,377 @@
+          this.velocities = spec.velocities;
+          this.initialized = true;
+        });
+  }
+
+  /**
+   * Map pitch and velocity to sample name.
+   */
+  private sampleInfoToName(sampleInfo: SampleInfo) {
+    if (this.velocities) {
+      return `p${sampleInfo.pitch}_v${sampleInfo.velocity}`;
+    } else {
+      return `p${sampleInfo.pitch}`;
+    }
+  }
+
+  /**
+   * Map pitch and velocity to sample URL.
+   */
+  private sampleInfoToURL(sampleInfo: SampleInfo) {
+    return `${this.baseURL}/${this.sampleInfoToName(sampleInfo)}.mp3`;
+  }
+
+  /**
+   * Find nearest sampled velocity to a target velocity.
+   */
+  private nearestVelocity(velocity: number): number {
+    if (!this.velocities) {
+      return velocity;
+    }
+
+    let bestVelocity = undefined;
+    let bestDistance = constants.MIDI_VELOCITIES;
+    this.velocities.forEach((v) => {
+      const d = Math.abs(v - velocity);
+      if (d < bestDistance) {
+        bestVelocity = v;
+        bestDistance = d;
+      }
+    });
+    return bestVelocity;
+  }
+
+  /**
+   * Load samples necessary to play a set of pitch/velocity pairs. This must be
+   * called before any notes can be played.
+   *
+   * @param samples Array of pitch/velocity pairs.
+   */
+  async loadSamples(samples: SampleInfo[]) {
+    if (!this.initialized) {
+      await this.initialize();
+    }
+
+    const nearestSamples = Array.from(new Set(
+        samples
+            .filter((info) => {
+              if (info.pitch < this.minPitch || info.pitch > this.maxPitch) {
+                console.log(
+                    `Pitch ${info.pitch} is outside the valid range for ${
+                        this.name}, ignoring.`);
+                return false;
+              } else {
+                return true;
+              }
+            })
+            .map((info) => ({
+                   pitch: info.pitch,
+                   velocity: this.nearestVelocity(info.velocity)
+                 }))));
+
+    const sampleNamesAndURLs =
+        nearestSamples
+            .map((info) => ({
+                   name: this.sampleInfoToName(info),
+                   url: this.sampleInfoToURL(info)
+                 }))
+            .filter((nameAndURL) => !this.buffers.has(nameAndURL.name));
+
+    sampleNamesAndURLs.forEach(
+        (nameAndURL) => this.buffers.add(nameAndURL.name, nameAndURL.url));
+
+    await new Promise(resolve => Tone.Buffer.on('load', resolve));
+
+    sampleNamesAndURLs.forEach(
+        (nameAndURL) =>
+            console.log(`Loaded ${nameAndURL.name} for ${this.name}.`));
+  }
+
+  /**
+   * Play a note using one of the samples.
+   *
+   * @param pitch Pitch of the note.
+   * @param velocity Velocity of the note.
+   * @param startTime Time at which to start playing the note.
+   * @param duration Length of the note in seconds.
+   */
+  playNote(
+      pitch: number, velocity: number, startTime: number, duration: number) {
+    if (!this.initialized) {
```

Comment:
why not just initialize this lazily?

Reply:
Like above, implication is that playNote schedules immediately.